src/anyp/TrafficMode.h

```diff
@@ -72,9 +87,61 @@ class TrafficModeFlags
      *  - encrypted outbound server connections
      *  - peer relay prohibited. TODO: re-encrypt and re-wrap with CONNECT
      */
-    bool tunnelSslBumping = false;
+    bool tunnelSslBumping;
+};
 
-    PortKind portKind; ///< the parsed port type value
+inline std::ostream &
+operator <<(std::ostream &os, const TrafficModeFlags &flags)
+{
+    const char *kindStr = nullptr;
+    if (flags.portKind == TrafficModeFlags::httpPort)
+        kindStr = "http_port";
+    else if (flags.portKind == TrafficModeFlags::httpsPort)
+        kindStr = "https_port";
+    else {
+        assert(flags.portKind == TrafficModeFlags::ftpPort);
+        kindStr = "ftp_port";
+    }
+    return os << kindStr << " with accel: " << flags.accelSurrogate <<
+        "require-proxy-header: " << flags.proxySurrogateHttp <<
+        "intercept: " << flags.natIntercept <<
+        "tproxy: " << flags.tproxyIntercept <<
+        "ssl-bump: " << flags.tunnelSslBumping;
+}
+
+constexpr std::array<TrafficModeFlags, 18> AcceptableTrafficModeFlags =
+{
+    {
+        // portKind; accelSurrogate, proxySurrogateHttp, natIntercept, tproxyIntercept, tunnelSslBumping
+
+        // accel
+        {TrafficModeFlags::httpPort, true, false, false, false, false},
+        {TrafficModeFlags::httpPort, true, false, true, false, false},
+        {TrafficModeFlags::httpPort, true, false, true, false, true},
+
+        // natIntercept
+        {TrafficModeFlags::httpPort, false, false, true, false, false},
+        {TrafficModeFlags::httpPort, false, false, true, false, true},
+        {TrafficModeFlags::httpPort, false, true, true, false, false},
+        {TrafficModeFlags::httpPort, false, true, true, false, true},
+
+        // tproxyIntercept
+        {TrafficModeFlags::httpPort, false, false, false, true, false},
+        {TrafficModeFlags::httpPort, false, false, false, true, true},
+        {TrafficModeFlags::httpPort, false, true, false, true, false},
+        {TrafficModeFlags::httpPort, false, true, false, true, true},
+
+        // proxySurrogateHttp
+        {TrafficModeFlags::httpPort, false, true, false, false, false},
+
+        {TrafficModeFlags::httpsPort, true, false, false, false, false},
+        {TrafficModeFlags::httpsPort, false, true, false, false, true},
+        {TrafficModeFlags::httpsPort, false, false, true, false, true},
+        {TrafficModeFlags::httpsPort, false, false, false, true, true},
+
+        {TrafficModeFlags::ftpPort, false, false, true, false, false},
+        {TrafficModeFlags::ftpPort, false, false, false, true, false}
+    }
```

Comment:
I assume the following supported cases are missing from code or from your summary table. They are considered "supported" or "allowed" in the snippet below:

```C++
{TrafficModeFlags::httpPort, false, false, false, false, false},
{TrafficModeFlags::httpPort, false, false, false, false, true},
{TrafficModeFlags::httpPort, false, true, false, false, true},
```

And here is how the whole table might look if we use `if` statements with helper functions:

```C++
if (flags.portKind == TrafficModeFlags::httpPort) {
    if (flags.accelSurrogate) {

        // {TrafficModeFlags::httpPort, true, false, false, false, false},
        // {TrafficModeFlags::httpPort, true, false, true, false, false},
        // {TrafficModeFlags::httpPort, true, false, true, false, true},

        cfg.rejectFlag("because it is not yet supported on http_port accel",
            &TrafficModeFlags::proxySurrogateHttp);
        cfg.rejectFlag("because TODO: why not?",
            &TrafficModeFlags::tproxyIntercept);
        // TODO: Why does tunnelSslBumping imply (in boolean logic sense) natIntercept?!
        cfg.checkImplication(&TrafficModeFlags::tunnelSslBumping,
            &TrafficModeFlags::natIntercept);
        return;
    }

    // {TrafficModeFlags::httpPort, false, false, false, false, *},
    // {TrafficModeFlags::httpPort, false, false, true, false, *},
    // {TrafficModeFlags::httpPort, false, false, false, true, *},
    // {TrafficModeFlags::httpPort, false, true, false, false, *},
    // {TrafficModeFlags::httpPort, false, true, true, false, *},
    // {TrafficModeFlags::httpPort, false, true, false, true, *},
    cfg.allowEither(
        &TrafficModeFlags::natIntercept,
        &TrafficModeFlags::tproxyIntercept);
    return;
}

if (flags.portKind == TrafficModeFlags::httpsPort) {
    // {TrafficModeFlags::httpsPort, true, false, false, false, false},
    // {TrafficModeFlags::httpsPort, false, true, false, false, true},
    // {TrafficModeFlags::httpsPort, false, false, true, false, true},
    // {TrafficModeFlags::httpsPort, false, false, false, true, true},

    if (flags.accelSurrogate) {
        cfg.rejectFlag("because it is not supported by https_port accel",
            &TrafficModeFlags::proxySurrogateHttp,
            &TrafficModeFlags::natIntercept,
            &TrafficModeFlags::tproxyIntercept,
            &TrafficModeFlags::tunnelSslBumping);
    } else {
        // XXX: https_port does not really _require_ acceleration. Without
        // accelSurrogate, the port works as an HTTPS forward proxy. This
        // condition should be removed AFAICT, but more refactoring is needed to
        // support basic HTTPS forward proxy (without any other flags):
        // {TrafficModeFlags::httpsPort, false, false, false, false, false},
        cfg.requireAll(
            &TrafficModeFlags::tunnelSslBumping);

        cfg.requireEither(
            &TrafficModeFlags::proxySurrogateHttp,
            &TrafficModeFlags::natIntercept,
            &TrafficModeFlags::tproxyIntercept);
    }
    return;
}

if (flags.portKind == TrafficModeFlags::ftpPort) {

    // {TrafficModeFlags::ftpPort, false, false, true, false, false},
    // {TrafficModeFlags::ftpPort, false, false, false, true, false}

    cfg.rejectFlag("because it is not supported by ftp_port",
        &TrafficModeFlags::accelSurrogate,
        &TrafficModeFlags::proxySurrogateHttp,
        &TrafficModeFlags::tunnelSslBumping);

    // XXX: IIRC, ftp_port does not require interception of any kind so
    // this may need to be changed from requireEither() to allowEither(), adding
    // {TrafficModeFlags::ftpPort, false, false, false, false, false}
    cfg.requireEither(
        &TrafficModeFlags::natIntercept,
        &TrafficModeFlags::tproxyIntercept);
    return;
}
```

... where

* rejectFlag() rejects flag combinations where any of the given flags is set.
* allowEither() rejects flag combinations where more than one flag is set.
* requireEither() is allowEither() that also rejects flag combinations where no flag is set
* requireAll() rejects flag combinations where at least one flag is unset.
* checkImplication(a,b) rejects the `a && !b` combination only

The above sketch still needs more work (see XXXs and TODOs), but I hope it gives you an overall idea on how this can be refactored to become more friendly to developers and admins. The ball is on your side now.

Reply:
I implemented this sketch (up to f09fd6f). Note that my initial table (which this sketch was based on) had some bugs, so the conditions(based on the old code) proved to be a little different. With the corrected version, I think that the proposed XXXs and TODOs became irrelevant.